lib/services/watcher.go

```diff
@@ -1666,3 +1666,159 @@ func (p *accessRequestCollector) processEventAndUpdateCurrent(ctx context.Contex
+	}
+	if cfg.SAMLIdPServiceProvidersC == nil {
+		cfg.SAMLIdPServiceProvidersC = make(chan types.SAMLIdPServiceProviders)
+	}
+	return nil
+}
+
+// NewSAMLIdPServiceProviderWatcher returns a new instance of SAMLIdPServiceProviderWatcher.
+func NewSAMLIdPServiceProviderWatcher(ctx context.Context, cfg SAMLIdPServiceProviderWatcherConfig) (*SAMLIdPServiceProviderWatcher, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	collector := &samlIDPServiceProviderCollector{
+		SAMLIdPServiceProviderWatcherConfig: cfg,
+		initializationC:                     make(chan struct{}),
+	}
+	watcher, err := newResourceWatcher(ctx, collector, cfg.ResourceWatcherConfig)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &SAMLIdPServiceProviderWatcher{watcher, collector}, nil
+}
+
+// SAMLIdPServiceProviderWatcher is built on top of resourceWatcher to monitor SAML IdP service provider resources.
+type SAMLIdPServiceProviderWatcher struct {
+	*resourceWatcher
+	*samlIDPServiceProviderCollector
+}
+
+// samlIDPServiceProviderCollector accompanies resourceWatcher when monitoring SAML IdP service provider resources.
+type samlIDPServiceProviderCollector struct {
+	// SAMLIdPServiceProviderWatcherConfig is the watcher configuration.
+	SAMLIdPServiceProviderWatcherConfig
+	// current holds a map of the currently known SAML IdP service provider resources.
+	current map[string]types.SAMLIdPServiceProvider
+	// lock protects the "current" map.
+	lock sync.RWMutex
+	// initializationC is used to check that the watcher has been initialized properly.
+	initializationC chan struct{}
+	once            sync.Once
+}
+
+// resourceKind specifies the resource kind to watch.
+func (p *samlIDPServiceProviderCollector) resourceKind() string {
+	return types.KindSAMLIdPServiceProvider
+}
+
+// initializationChan is used to check if the initial state sync has been completed.
+func (p *samlIDPServiceProviderCollector) initializationChan() <-chan struct{} {
+	return p.initializationC
+}
+
+// getResourcesAndUpdateCurrent refreshes the list of current resources.
+func (p *samlIDPServiceProviderCollector) getResourcesAndUpdateCurrent(ctx context.Context) error {
+	var samlIDPServiceProviders []types.SAMLIdPServiceProvider
+	nextToken := ""
+	for {
+		var sps []types.SAMLIdPServiceProvider
+		var err error
+		sps, nextToken, err = p.SAMLIdPServiceProviders.ListSAMLIdPServiceProviders(ctx, p.PageSize, nextToken)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		samlIDPServiceProviders = append(samlIDPServiceProviders, sps...)
+		if nextToken == "" {
+			break
+		}
+	}
+
+	newCurrent := make(map[string]types.SAMLIdPServiceProvider, len(samlIDPServiceProviders))
+	for _, samlIDPServiceProvider := range samlIDPServiceProviders {
+		newCurrent[samlIDPServiceProvider.GetName()] = samlIDPServiceProvider
+	}
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	p.current = newCurrent
+	p.defineCollectorAsInitialized()
+
+	select {
```

Comment:
Do we need to hold the lock for the duration of the select?

Reply:
No, this should be okay here. The order in which these objects are sent to the channel shouldn't matter.